detector/calorimeter/HCalThreePartsEndcap_o1_v01_geo.cpp

```diff
@@ -1,467 +1,496 @@
-
-  unsigned int numSequencesR1 = 0;
-  unsigned int numSequencesR2 = 0;
-  unsigned int numSequencesR3 = 0;
-  double moduleDepth1 = 0.;
-  double moduleDepth2 = 0.;
-  double moduleDepth3 = 0.;
-  std::vector<double> layerDepths1 = std::vector<double>();
-  std::vector<double> layerDepths2 = std::vector<double>();
-  std::vector<double> layerDepths3 = std::vector<double>();
-
-  // get all 'layer' children of the 'layers' tag
-  std::vector<xml_comp_t> Layers;
-  for (xml_coll_t xCompColl(aXmlElement.child(_Unicode(layers)), _Unicode(layer)); xCompColl;
-       ++xCompColl) {
-    Layers.push_back(xCompColl);
-  }
-  
-  for (std::vector<xml_comp_t>::iterator it = Layers.begin(); it != Layers.end(); ++it) {
-    xml_comp_t layer = *it;
-    Dimension layerDimension(layer.dimensions());
-    numSequencesR1 += layerDimension.nmodules();
-    numSequencesR2 += layerDimension.nModules();
-    numSequencesR3 += layerDimension.nPads();
-    for (int nLayer = 0; nLayer < layerDimension.nmodules(); nLayer++) {
-      moduleDepth1 += layerDimension.dr();
-      layerDepths1.push_back(layerDimension.dr());
-    }
-    for (int nLayer = 0; nLayer < layerDimension.nModules(); nLayer++) {
-      moduleDepth2 += layerDimension.dr();
-      layerDepths2.push_back(layerDimension.dr());
-    }
-    for (int nLayer = 0; nLayer < layerDimension.nPads(); nLayer++) {
-      moduleDepth3 += layerDimension.dr();
-      layerDepths3.push_back(layerDimension.dr());
+static dd4hep::Ref_t createHCalEC(dd4hep::Detector& lcdd, xml_h xmlElement, dd4hep::SensitiveDetector sensDet){
+    xml_det_t xmlDet = xmlElement;
+    std::string detName = xmlDet.nameStr();
+
+    // Make DetElement
+    dd4hep::DetElement caloDetElem(detName, xmlDet.id());
+
+    // Make volume that envelopes the whole barrel; set material to air
+    Dimension dimensions(xmlDet.dimensions());
+
+    dd4hep::Tube envelope(dimensions.rmin(), dimensions.rmax1(), (dimensions.v_offset() + dimensions.z_length()));
+    dd4hep::Tube negative1(dimensions.rmin(), dimensions.rmax1(), (dimensions.offset() - dimensions.width()));
+    dd4hep::Tube negative2(dimensions.rmin(), dimensions.rmin1(), (dimensions.z_offset() - dimensions.dz()));
+    dd4hep::Tube negative3(dimensions.rmin(), dimensions.rmin2(), (dimensions.v_offset() - dimensions.z_length()));
+
+    dd4hep::SubtractionSolid envelopeShapeTmp1(envelope, negative1);
+    dd4hep::SubtractionSolid envelopeShapeTmp2(envelopeShapeTmp1, negative2);
+    dd4hep::SubtractionSolid envelopeShape(envelopeShapeTmp2, negative3);
+
+    Volume envelopeVolume(detName + "_volume", envelopeShape, lcdd.air());
+    envelopeVolume.setVisAttributes(lcdd, dimensions.visStr());
+
+    // Set sensitive detector type
+    Dimension sensDetType = xmlElement.child(_Unicode(sensitive));
+    sensDet.setType(sensDetType.typeStr());
+
+    // Dimension dimensions(xmlElement.child(_Unicode(dimensions)));
```

Comment:
remove commented line
possibly add a comment to explain meaning of retrieved dimensions in following lines

Reply:
done